analyzer/tools/build-logger/tests/unit/__init__.py

```diff
@@ -0,0 +1,219 @@
+    char = "\r"
+
+
+# The following 3 characters are below 0x20, so they need special treat.
+# However, these are not *explicitly* escaped by the ldlogger.
+class ack:
+    char = "\x06"
+
+
+class shiftin:
+    char = "\x0F"
+
+
+class datalinkescape:
+    char = "\x10"
+
+
+class escape:
+    char = "\x1B"
+
+
+class formfeed:
+    char = "\f"
+
+
+def concat(*args):
+    res = ""
+    for arg in args:
+        ntimes = 1
+        if type(arg) == tuple and len(arg) == 2:
+            ntimes = arg[1]
+            arg = arg[0]
+        char = getattr(arg, "char", None)
+        if type(arg) == str:
+            res += arg
+            continue
+        if not char:
+            raise KeyError()
+        res += char * ntimes
+    return res
+
+
+def _run_command(
+    cmd: Sequence[str],
+    cwd: Optional[str] = None,
+    env: Optional[Mapping[str, str]] = None,
+) -> Tuple[int, str, str]:
+    cmd = " ".join([shlex.quote(c) for c in cmd])
+    try:
+        proc = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            cwd=cwd,
+            env=env,
+            shell=True,
+            encoding="utf-8",
+            universal_newlines=True,
+            errors="ignore",
+        )
+        stdout, stderr = proc.communicate()
+        retcode = proc.returncode
+    except FileNotFoundError:
+        return 2, "", ""
+    return retcode, stdout, stderr
+
+
+_PRINT_STRINGIFIED_VARIABLE_CODE = """
+#include <stdio.h>
+#define xstr(a) str(a)
+#define str(a) #a
+int main() {
+  char buf[] = xstr(VARIABLE); // It might containt null characters.
+  printf("--");
+  for (unsigned i = 0; sizeof(buf) > 1 && i < sizeof(buf) - 1; ++i)
+    printf("%c", buf[i]);
+  printf("--");
+}
+"""
+
+empty_env = {"PATH": "/usr/bin"}
+
+
+class BasicLoggerTest(unittest.TestCase):
+    logger_dir = LOGGER_DIR
+
+    def setUp(self):
+        self.maxDiff = None
+        fd, self.source_file = tempfile.mkstemp(
+            suffix=".cpp", prefix="logger-test-source-", text=True
+        )
+        os.write(fd, bytes(_PRINT_STRINGIFIED_VARIABLE_CODE, "utf-8"))
+        os.close(fd)
+        fd, self.logger_file = tempfile.mkstemp(
+            suffix=".json", prefix="logger-test-db-", text=True
+        )
+        os.close(fd)
+        fd, self.binary_file = tempfile.mkstemp(
+            suffix=".out", prefix="logger-test-binary-"
+        )
+        os.close(fd)
+        fd, self.logger_debug_file = tempfile.mkstemp(
+            suffix=".log", prefix="logger-test-debug-", text=True
+        )
+        os.close(fd)
+
+    def tearDown(self):
+        tmp_files = [
+            self.source_file,
+            self.logger_file,
+            self.binary_file,
+            self.logger_debug_file,
+            self.logger_debug_file + ".lock",
+        ]
+        for file in tmp_files:
+            if os.path.isfile(file):
+                os.remove(file)
```

Comment:
Perhaps we should handle the case when `remove` fails?

Reply:
Now I'm using try-except.